### ISSUE TITLE
ASN1 Library

### ISSUE THREAD

**Issue by mwcw** [bcgit/pc-dart#35]
Hi,

Ultimately we need to be able to support ASN1 encoded messages as CMS, X.509 and others.

We need an ASN1 library so we do not force dependencies on out users.

MW

**Comment by Ephenodrom** [comment]
Why not use the existing asn1lib package? https://pub.dev/packages/asn1lib

I use this package in combination with the pointycastle package for my X509Utils class and it works really well.

I also got upload access for pub.dev so new features / bugfixes can be available very fast.

**Comment by mwcw** [comment]
We can't force dependencies on people, crypto libraries become infrastructure.

**Comment by AKushWarrior** [comment]
> We can't force dependencies on people, crypto libraries become infrastructure.

Fork it and package it with this library?

**Comment by Ephenodrom** [comment]
Any plans / roadmap for this? Should we start with listing all ASN1 objects that are covered in the java bouncycastle?

**Comment by Ephenodrom** [comment]
@mwcw I started to work on this, if this is ok! 

I used the ASN1Lib package as a template.  so that users of this package could easily switch to Pointycastle in the future without changing much code.

**Comment by mwcw** [comment]
> @mwcw I started to work on this, if this is ok!
> 
> I used the ASN1Lib package as a template. so that users of this package could easily switch to Pointycastle in the future without changing much code.

Hi,

Yes thank you very much for picking this up I agree with the strategy of easy switching as well.

MW

**Comment by Ephenodrom** [comment]
A small update on the current status :

The table lists all objects currently supported with the different encoding types. 

| ASN1 Type  | BER  | DER  | DL  |
|---|---|---|---|
| BOOLEAN  |   | X |   |
| INTEGER  | Only supports DER | X |   |
| BIT_STRING  |   | X |   |
| OCTET_STRING  |   | X  |   |
| NULL  |   | X |   |
| OBJECT_IDENTIFIER  | Only supports DER | X |   |
| ENUMERATED  |   | X |   |
| UTF8_STRING  |   | X |   |
| SEQUENCE  |   | X |   |
| SET  |   | X |   |
| PRINTABLE_STRING  |   | X |   |
| IA5_STRING  |   | X |   |
| UTC_TIME  |   | X |   |

**Comment by penhorwood** [comment]
Looking forward to this being roled into the package.  I need ASN.1 to handle some interactions with a server running java.

**Comment by penhorwood** [comment]
I know this is a open source project that is worked as we get the time.  I am waiting on the ASN.1 pull request by @Ephenodrom to continue a project I am working on.  Is it possible to get an ETA or something when this will be merged into the project?

**Comment by Ephenodrom** [comment]
The PR is merged and this is the state of the ASN1 primitives.

| ASN1 Type  | Decode From  | Encode To  |
|---|---|---|
| BOOLEAN  | * DER :heavy_check_mark:<br> | * DER :heavy_check_mark:<br> |
| INTEGER  | * DER :heavy_check_mark:<br> | * DER :heavy_check_mark:<br> |
| BIT_STRING  | * DER :heavy_check_mark:<br>* BER Constructed:heavy_check_mark:<br>* BER Long Length Form:heavy_check_mark:<br>* BER Padded:heavy_check_mark:<br>* BER Constructed Indefinite Length:heavy_check_mark: | * DER :heavy_check_mark:<br>* BER Constructed:heavy_check_mark:<br>* BER Long Length Form:heavy_check_mark:<br>* BER Padded:heavy_check_mark:<br>* BER Constructed Indefinite Length:heavy_check_mark: |
| OCTET_STRING  | * DER :heavy_check_mark:<br>* BER Constructed :heavy_check_mark:<br>* BER Long Length Form :heavy_check_mark:<br>* BER Constructed Indefinite Length :heavy_check_mark: | * DER :heavy_check_mark:<br>* BER Constructed :heavy_check_mark:<br>* BER Long Length Form :heavy_check_mark:<br>* BER Constructed Indefinite Length :heavy_check_mark:  |
| NULL  | * DER :heavy_check_mark:<br>* BER Long Length Form:heavy_check_mark:  | * DER :heavy_check_mark:<br>* BER Long Length Form:heavy_check_mark: |
| OBJECT_IDENTIFIER  | * DER :heavy_check_mark:<br> | * DER :heavy_check_mark:<br> |
| ENUMERATED  | * DER :heavy_check_mark:<br> | * DER :heavy_check_mark:<br> |
| UTF8_STRING |  * DER :heavy_check_mark:<br>* BER Constructed:heavy_check_mark:<br>* BER Long Length Form:heavy_check_mark:<br>* BER Constructed Indefinite Length:heavy_check_mark: | * DER :heavy_check_mark:<br>* BER Constructed:heavy_check_mark:<br>* BER Long Length Form:heavy_check_mark:<br>* BER Constructed Indefinite Length:heavy_check_mark:  |
| SEQUENCE  | * DER :heavy_check_mark:<br> | * DER :heavy_check_mark:<br> |
| SET  | * DER :heavy_check_mark:<br> | * DER :heavy_check_mark:<br>|
| PRINTABLE_STRING  | * DER :heavy_check_mark:<br>* BER Constructed:heavy_check_mark:<br>* BER Long Length Form:heavy_check_mark:<br>* BER Constructed Indefinite Length:heavy_check_mark: | * DER :heavy_check_mark:<br>* BER Constructed:heavy_check_mark:<br>* BER Long Length Form:heavy_check_mark:<br>* BER Constructed Indefinite Length:heavy_check_mark: |
| IA5_STRING  | * DER :heavy_check_mark:<br>* BER Constructed:heavy_check_mark:<br>* BER Long Length Form:heavy_check_mark:<br>* BER Constructed Indefinite Length:heavy_check_mark: | * DER :heavy_check_mark:<br>* BER Constructed:heavy_check_mark:<br>* BER Long Length Form:heavy_check_mark:<br>* BER Constructed Indefinite Length:heavy_check_mark:   |
| UTC_TIME  | * DER :heavy_check_mark:<br> | * DER :heavy_check_mark:<br> |

**Comment by AKushWarrior** [comment]
Hasn't this been merged? @mwcw can you close the issue?

**Comment by mwcw** [comment]
> Hasn't this been merged? @mwcw can you close the issue?

Yep! In the release!

Thanks again!
